### DISCUSSION TITLE
Type "reactive[None]" cannot be assigned to type

### DISCUSSION THREAD

**Discussion by davetapley** [Textualize/textual#1977] [Q&A] [answered]
The [docs here](https://textual.textualize.io/guide/reactivity/#typing-reactive-attributes) don't _explicitly_ show it, but I _presume_ this is should be valid:
```py
name: reactive[str | None] = reactive(None)
```

But in VSCode / Pylance I get:
```
Expression of type "reactive[None]" cannot be assigned to declared type "reactive[str | None]"
  Type "reactive[None]" cannot be assigned to type "ReactiveType@Reactive"
```

---

I initially tried:

```py
name: reactive[str | None] = reactive()
```

But that causes:

```
Argument missing for parameter "default"
```

---

```
# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.12.1 |
| Rich    | 13.3.1 |

## Python

| Name           | Value                                                    |
|----------------|----------------------------------------------------------|
| Version        | 3.10.9                                                   |
| Implementation | CPython                                                  |
| Compiler       | GCC 9.4.0                                                |
| Executable     | /home/dave/.pyenv/versions/3.10.9/envs/ng/bin/python3.10 |

## Operating System

| Name    | Value                                       |
|---------|---------------------------------------------|
| System  | Linux                                       |
| Release | 5.14.0-1034-oem                             |
| Version | #37-Ubuntu SMP Tue Apr 19 14:13:20 UTC 2022 |

## Terminal

| Name                 | Value           |
|----------------------|-----------------|
| Terminal Application | vscode (1.76.0) |
| TERM                 | xterm-256color  |
| COLORTERM            | truecolor       |
| FORCE_COLOR          | *Not set*       |
| NO_COLOR             | *Not set*       |

## Rich Console options

| Name           | Value               |
|----------------|---------------------|
| size           | width=68, height=59 |
| legacy_windows | False               |
| min_width      | 1                   |
| max_width      | 68                  |
| is_terminal    | True                |
| encoding       | utf-8               |
| max_height     | 59                  |
| justify        | None                |
| overflow       | None                |
| no_wrap        | False               |
| highlight      | None                |
| markup         | None                |
| height         | None                |
```

**Comment by davep** [comment, marked as answer]
As best as I've been able to tell, this is an issue with pyright/pylance. I see it too when using pyright in Emacs. I'm finding I have to give pyright a helping hand like this:

```python
name: reactive[str | None] = reactive[str | None](None)
```

or, to be compatible with some older supported Pythons:

```python
name: reactive[str | None] = reactive[Optional[str]](None)
```

(assuming `Optional` is imported from `typing`).

**Reply by davetapley** [reply]
That workaround ✅ 

It'd be nice to have a Pyright issue to link to.
I just spent a bit of time ripping out the relevant parts of `reactive.py` and `repr.py` to try and get a repro, however:

1. If I put it [all in one file](https://gist.github.com/davetapley/9bbe94aee6a6bbd1386c16dca3b962ec) it type checks just fine 😑 
2. I tried [separating out `repr.py`](https://gist.github.com/davetapley/0fd61bb00dadfa2eeb05a5f5a51bf401), but that causes another error: `Expected no type arguments for class "reactive"`.

Right up on the limit of my comfort with Python here though, so I'm not sure how to proceed.

I assume the `Callable` shows up in `repr.py` because [this](https://stackoverflow.com/a/65937450/21115)?
If so I wonder if something PEP 612 would help with, but again it'd be nice to have a clean repro to check against.

**Reply by TomJGooding** [reply]
I'm very much out of my depth here, but it looks like you haven't actually imported `repr.py` in your separated example...?

**Reply by davetapley** [reply]
@TomJGooding oh, yes, sorry, copy/paste error!

**Reply by davetapley** [reply]
I've run in to this again and opened:
- https://github.com/microsoft/pylance-release/issues/4656
